arrow-fx-stm/src/main/kotlin/arrow/fx/stm/TMap.kt

```diff
@@ -8,6 +8,153 @@ fun <K, V> STM.newTMap(fn: (K) -> Int): TMap<K, V> = TMap(newHamt(), fn)
 fun <K, V> STM.newTMap(): TMap<K, V> = newTMap { it.hashCode() }
 fun <K, V> STM.newTMap(hash: Hash<K>): TMap<K, V> = newTMap { hash.run { it.hash() } }
 
+/**
+ * A [TMap] is a concurrent transactional implementation of a key value hashmap.
+ *
+ * Based on a Hash-Array-Mapped-Trie implementation. While this does mean that a read may take up to 5 steps to be resolved (depending on how
+ *  well distributed the hash function is), it also means that structural changes can be isolated and thus do not increase contention with
+ *  other transactions. This effectively means concurrent access to different values is unlikely to interfere with each other.
+ *
+ * > Hash conflicts are resolved by chaining.
```

Comment:
What does it mean by chaining?

Reply:
There are two ways to to resolve hash conflicts:
- Either replace the value and assume hashes are unique
- Or put them in a collection and iterate through them on lookups

Chaining here refers to chaining it to the end of the leaf array, I've heard and seen it used in this context while I did a bit of research on how Hamt structures work so I just used it here.